wallet/wallet-cli-lib/src/repl/interactive/mod.rs

```diff
@@ -0,0 +1,182 @@
+        )))
+        .with_menu(ReedlineMenu::HistoryMenu(Box::new(
+            ListMenu::default().with_name(HISTORY_MENU_NAME),
+        )));
+
+    let edit_mode: Box<dyn EditMode> = if vi_mode {
+        let mut normal_keybindings = default_vi_normal_keybindings();
+        let mut insert_keybindings = default_vi_insert_keybindings();
+
+        add_menu_keybindings(&mut normal_keybindings);
+        add_menu_keybindings(&mut insert_keybindings);
+
+        Box::new(Vi::new(insert_keybindings, normal_keybindings))
+    } else {
+        let mut keybindings = default_emacs_keybindings();
+        add_menu_keybindings(&mut keybindings);
+
+        Box::new(Emacs::new(keybindings))
+    };
+
+    line_editor = line_editor.with_edit_mode(edit_mode);
+
+    Ok(line_editor)
+}
+
+fn process_line(
+    repl_command: &Command,
+    event_tx: &mpsc::UnboundedSender<Event>,
+    sig: reedline::Signal,
+) -> Result<Option<ConsoleCommand>, WalletCliError> {
+    let line = match sig {
+        Signal::Success(line) => line,
+        Signal::CtrlC => {
+            // Prompt has been cleared and should start on the next line
+            return Ok(None);
+        }
+        Signal::CtrlD => {
+            return Ok(Some(ConsoleCommand::Exit));
+        }
+    };
+
+    let command_opt = parse_input(&line, repl_command)?;
+
+    let command = match command_opt {
+        Some(command) => command,
+        None => return Ok(None),
+    };
+
+    super::run_command_blocking(event_tx, command).map(Option::Some)
+}
+
+pub fn run(
+    output: &ConsoleContext,
+    event_tx: mpsc::UnboundedSender<Event>,
+    exit_on_error: bool,
+    printer: reedline::ExternalPrinter<String>,
+    history_file: Option<PathBuf>,
+    vi_mode: bool,
+) -> Result<(), WalletCliError> {
+    let repl_command = get_repl_command();
+    let mut line_editor =
+        create_line_editor(printer, repl_command.clone(), output, history_file, vi_mode)?;
+
+    let prompt = wallet_prompt::WalletPrompt::new();
+
+    loop {
+        let sig = line_editor.read_line(&prompt).expect("Should not fail normally");
+
+        let res = process_line(&repl_command, &event_tx, sig);
+
+        match res {
+            Ok(Some(ConsoleCommand::Print(text))) => {
+                cli_println!(output, "{}", text);
+            }
+            Ok(Some(ConsoleCommand::ClearScreen)) => {
+                line_editor.clear_scrollback().expect("Should not fail normally");
+            }
+            Ok(Some(ConsoleCommand::ClearHistory)) => {
+                line_editor.history_mut().clear().expect("Should not fail normally");
+            }
+            Ok(Some(ConsoleCommand::PrintHistory)) => {
+                line_editor.print_history().expect("Should not fail normally");
+            }
+            Ok(Some(ConsoleCommand::Exit)) => return Ok(()),
+
+            Ok(None) => {}
+
+            Err(err) => {
+                if exit_on_error {
+                    return Err(err);
+                }
```

Comment:
Should the error be printed out even if `exit_on_error` is enabled? Or does it happen somewhere else?

Reply:
Yes, the error is printed from main in this case.